File: examples/bakery-interaction-examples/reserve-items/src/main/scala/webshop/webservice/ReserveItems.scala
package webshop.webservice

import com.ing.baker.recipe.annotations.FiresEvent

import scala.concurrent.Future
import com.ing.baker.recipe.javadsl.Interaction

trait ReserveItems extends Interaction {

  case class Item(itemId: String)

  case class OrderPlaced(Items: List[Item])

  case class ReservedItems(items: List[Item], data: Array[Byte])

  sealed trait ReserveItemsOutput

  case class OrderHadUnavailableItems(unavailableItems: List[Item]) extends ReserveItemsOutput

  case class ItemsReserved(reservedItems: ReservedItems) extends ReserveItemsOutput

  @FiresEvent(oneOf = Array(classOf[ItemsReserved], classOf[OrderHadUnavailableItems]))
  def apply(items: List[Item]): Future[ReserveItemsOutput]
}

File: examples/baker-example/src/main/java/webshop/simple/ReserveItems.java
package webshop.simple;

import com.ing.baker.recipe.annotations.FiresEvent;
import com.ing.baker.recipe.annotations.RequiresIngredient;
import com.ing.baker.recipe.javadsl.Interaction;
import webshop.simple.ingredients.Item;
import webshop.simple.ingredients.ReservedItems;

import java.util.List;

public interface ReserveItems extends Interaction {

    interface ReserveItemsOutcome {
    }

    class OrderHadUnavailableItems implements ReserveItemsOutcome {

        public final List<Item> unavailableItems;

        public OrderHadUnavailableItems(List<Item> unavailableItems) {
            this.unavailableItems = unavailableItems;
        }
    }

    class ItemsReserved implements ReserveItemsOutcome {

        public final ReservedItems reservedItems;

        public ItemsReserved(ReservedItems reservedItems) {
            this.reservedItems = reservedItems;
        }
    }

    @FiresEvent(oneOf = {OrderHadUnavailableItems.class, ItemsReserved.class})
    ReserveItemsOutcome apply(@RequiresIngredient("items") List<Item> items);
}
